src/muc/mongoose_muc_online_cets.erl

```diff
@@ -0,0 +1,113 @@
+    cets:start(Tab, #{handle_conflict => fun ?MODULE:handle_conflict/2}),
+    cets_discovery:add_table(mongoose_cets_discovery, Tab),
+    ok.
+
+%% We should keep one room and stop another room
+%% But stopping logic needs to be tested heavily and designed
+%% because we would need to figure out how to send presences to participants
+%% (and maybe document how to rejoin the kicked room)
+-spec handle_conflict(muc_tuple(), muc_tuple()) -> muc_tuple().
+handle_conflict(Rec1, Rec2) when Rec1 > Rec2 ->
+    Rec1;
+handle_conflict(_Rec1, Rec2) ->
+    Rec2.
+
+-spec register_room(
+        HostType :: mongooseim:host_type(),
+        MucHost :: jid:lserver(),
+        Room :: mod_muc:room(),
+        Pid :: pid()) -> ok | {exists, pid()} | {error, term()}.
+register_room(HostType, MucHost, Room, Pid) ->
+    register_room(HostType, MucHost, Room, Pid, 3).
+
+register_room(_HostType, _MucHost, _Room, _Pid, 0) ->
+    {error, failed_to_register};
+register_room(HostType, MucHost, Room, Pid, Retries) ->
+    Tab = table_name(HostType),
+    Rec = {{MucHost, Room}, Pid},
+    case find_room_pid(HostType, MucHost, Room) of
+        {ok, OtherPid} ->
+            {exists, OtherPid};
+        {error, not_found} ->
+            case cets:insert_new(Tab, Rec) of
+                true ->
+                    ok;
+                false ->
+                     register_room(HostType, MucHost, Room, Pid, Retries - 1)
+            end
+    end.
+
+%% Race condition is possible between register and room_destroyed
+%% (Because register is outside of the room process)
+-spec room_destroyed(mongooseim:host_type(), jid:lserver(), mod_muc:room(), pid()) -> ok.
+room_destroyed(HostType, MucHost, Room, Pid) ->
+    Tab = table_name(HostType),
+    Rec = {{MucHost, Room}, Pid},
+    cets:delete_object(Tab, Rec),
+    ok.
+
+-spec find_room_pid(mongooseim:host_type(), jid:server(), mod_muc:room()) ->
+    {ok, pid()} | {error, not_found}.
+find_room_pid(HostType, MucHost, Room) ->
+    Tab = table_name(HostType),
+    case ets:lookup(Tab, {MucHost, Room}) of
+        [{_, Pid}] ->
+            {ok, Pid};
+        [] ->
+            {error, not_found}
+    end.
+
+%% This is used by MUC discovery but it is not very scalable.
+%% This function should look like get_online_rooms(HostType, MucHost, AfterRoomName, Limit)
+%% to reduce the load and still have pagination working.
+-spec get_online_rooms(mongooseim:host_type(), jid:lserver()) ->
+    [mod_muc:muc_online_room()].
+get_online_rooms(HostType, MucHost) ->
+    Tab = table_name(HostType),
+    [#muc_online_room{name_host = {Room, MucHost}, pid = Pid, host_type = HostType}
+     || [Room, Pid] <- ets:match(Tab, {{MucHost, '$1'}, '$2'})].
+
+-spec node_cleanup(mongooseim:host_type(), node()) -> ok.
+node_cleanup(HostType, Node) ->
+    Tab = table_name(HostType),
+    Pattern = {'_', '$1'},
+    Guard = {'==', {node, '$1'}, Node},
+    ets:select_delete(Tab, [{Pattern, [Guard], [true]}]),
```

Comment:
It's strange that I didn't come up with this question in one of the earlier reviews 😅 I understand that for lookups and reads we can use straight `ets` API but why is it safe to use it for deleting items? On the other hand in `room_destroyed` function there is a call for `cets` API - `cets:delete_object`. What's the difference?

Reply:
node_cleanup is called on **all** nodes in the cluster, so it is just data removal from the local table - no need for it to be replicated. And it is actually a lot of data removal to replicate, so we skip it.


Also, there is a chance we could move node_cleanup into CETS handle_down function - to avoid race conditions in node down detection. But it is another topic :)

Reply:
so, we skip replication step here, like it is done with mnesia backends

Reply:
Thanks